### ISSUE TITLE
tf-controller APM Support via OpenTelemetry

### ISSUE THREAD

**Issue by o6uoq** [weaveworks/tf-controller#413]
APM (Application Performance Monitoring) is a great way to provide insights into application-layer transactions (and more).

However, APMs are generally based on a language and/or framework e.g. Java, Go, Tomcat, Dropwizard, Ruby on Rails.

Due to this difference, not all APM agents and instrumentation are the same. For example, Java can instrument your application at run-time with automagic instrumentation out of the box. For Go, you have to instrument at compile-time (see below for more info) and instrument each transaction within source. Some agents provide a unified solution out of the box with minimum bootstrapping required (like Ruby).

By example and comparison, [Flagger](https://flagger.app/) has [Metrics Analysis](https://docs.flagger.app/usage/metrics) integrations with the likes of Prometheus, Datadog, Amazon CloudWatch, New Relic, et al. As `tf-controller` matures, then a nice to have will be to integrate with APMs.

Like above, not APM providers are cut the same. However, [OpenTracing](https://opentracing.io/) has now been superseded by [OpenTelemetry](https://opentelemetry.io/docs/migration/opentracing/), which opens up compatability to an even wider range of [APM providers](https://opentelemetry.io/vendors/). This _should_ mean we can integrate once, provide many times.

**TL;DR** - Intergrate `tf-controller` with OpenTelemetry to provide APM observability to a market wide range of providers.

**Comment by tomhuang12** [comment]
I agree that if we need to support APM, OpenTelemetry is the way to go. Some APM tools provide out of box APM tracing without manual instrumentations but understand that its tool to tool specific.
